### DISCUSSION TITLE
Synchroniously getting a part of output of a sub-process

### DISCUSSION THREAD

**Discussion by irvnriir** [dotnet/runtime#90882] [Q&A] [answered]
Maybe this doesn't work due to incorrect use of .NET functions, ~~but result of running this script is such a nonsense that i feel like its a bug in .NET~~ .

The `.ReadToEnd(` works here, but why to hell when im switching to `ReadBlock(` , this script outputs "10" and then gets stuck ? Obviously i don't want to run any additional operations for each output char, so aren't looking at `.Read(` . (the Stack Overflow is intentional and expected to be handled consistently)

Btw:
* using `.ReadLine(` locks the initial Process, i assume that: until a newline is output .
* using `.Read(` outputs "116" and locks the initial Process .

Also, in case .NET doesn't have this functionality, if async `.OutputDataReceived(` is used, how to determine that all output was processed ?

.NET Framework 4.8.1
```
// [req]
	using Console = System.Console;
	using Collections = System.Collections;
	using MethodImplAttribute = System.Runtime.CompilerServices.MethodImplAttribute;
	using MethodImplOptions = System.Runtime.CompilerServices.MethodImplOptions;
// 
abstract partial
class ac_DotNet {

public static int
Main (params string[] args) {
	
	if (args.Length == 0 || (args[0] != "1")) {
		string activeFile = System.Environment.GetCommandLineArgs()[0];
		System.Diagnostics.Process
		process = new System.Diagnostics.Process {
			StartInfo = new System.Diagnostics.ProcessStartInfo {
				FileName = activeFile,
				Arguments = "1",
				Verb = "",		// options can be received by using `.Verbs`
				CreateNoWindow = true,
				WindowStyle = System.Diagnostics.ProcessWindowStyle.Normal,
				UseShellExecute = false,		// run default Open action for the target instead of allowing only to start executables
				
				//# allows to manage the content by using Events and other interfaces
				RedirectStandardError = true,
				// RedirectStandardInput = true,
				RedirectStandardOutput = true
				
				// StandardErrorEncoding = te.Encoding.Utf8NoBom,
				// StandardOutputEncoding = te.Encoding.Utf8NoBom
			},
			EnableRaisingEvents = true,		// enables auto-running `.Exited`
			// SynchronizingObject = null
		};
		process.Exited += new System.EventHandler ((object subjectProcess, System.EventArgs g_param_f) => {
			Console.WriteLine(string.Join(Console.Out.NewLine, new string[] {
				"",
				"`.Exited`"
			}));
		});
		
		process.Start();
		
		int si1 = 10;		// `4096` , `42487`
		char[] te1 = new char[si1];
		Console.Write(process.StandardOutput.ReadBlock(te1, 0, si1));
		// Console.Write(process.StandardOutput.ReadToEnd());
		
		//# [solution]
		// int si1 = 10;		// `4096` , `42487`
		// char[] te1 = new char[si1];
		// while (!process.StandardOutput.EndOfStream) {
		// 	process.StandardOutput.ReadBlock(te1, 0, si1);
		// 	Console.Write(te1);
		// }
		
		process.WaitForExit();
		process.Refresh();
		Console.WriteLine(string.Join(Console.Out.NewLine, new string[] {
			"",
			string.Concat("`.ExitCode` : ", process.ExitCode.ToString(), " .")
		}));
		process.Dispose();
	} else {
		Console.WriteLine("t1");
		f1(int.Parse(args[0]));
	}
	
	Console.Write(string.Join(Console.Out.NewLine, new string[] {
		"",
		"Press Enter to continue..."
	}));
	Console.ReadLine();
	
	return 0;
}

[MethodImplAttribute(MethodImplOptions.NoInlining)]
static void
f1 (int v1) {
	System.Console.Write("t2");
	f1(v1);
}

}
```

**Comment by Clockwork-Muse** [comment]
If you're setting up two processes to communicate with each other, it's preferrable (for multiple reasons) to use something other than the console to do so.

Additionally, it would be beneficial to move to .NET Core if possible.

**Reply by irvnriir** [reply]
The sub-Process is for running a third-party console .exe .

**Comment by huoyaoyuan** [comment]
It looks like the process doesn't exit since it's stdout is blocked. Reading through the end of the pipe will unblock it and make it gracefully exit.

**Reply by irvnriir** [reply]
Apparently . But though idk why would it be blocked, because in total it even outputs much more than 10 chars which the `.ReadBlock(` requires . Im not stating that `.ReadBlock(` should work, but maybe anyone knows a way to achieve the goal, because i can't find anything in the documentation .

**Reply by huoyaoyuan** [reply, marked as answer]
> because in total it even outputs much more than 10 chars which the `.ReadBlock(` requires

It should be `WaitForExit` blocking. The process is blocked because the output buffer is full.

You can try to observe this by writing to stdout and stderr alternately, and see whether you can get output in the other stream.

**Reply by irvnriir** [reply]
This is exactly what i was looking for, ty alot
```
		int si1 = 10;		// `4096` , `42487`
		char[] te1 = new char[si1];
		while (!process.StandardOutput.EndOfStream) {
			process.StandardOutput.ReadBlock(te1, 0, si1);
			Console.Write(te1);
		}
```
